### PR TITLE
initial implementation of AMP live blog

### PR DESCRIPTION
**Pull request by SiAdcock**
## What does this change?

This is a first pass at implementing Live Blogs using AMP. The intention is to get the simplest example of live blogs into production. 

As it stands, the live update feature will not work. It will not be indexed by Google.
## What is the value of this and can you measure success?

As an MVP, this will provide a foundation upon which additional features (such as live updates) and design work can be built.
## Does this affect other platforms - Amp, Apps, etc?

Existing non-AMP live blogs will not be affected. Existing AMP articles will not be affected.
## Screenshots

picture 44
## Request for comment

@sndrs

### CODE REVIEW

File: article/app/views/liveblog/liveBlogBody.scala.html [hunk 1]
```diff
@@ -108,6 +107,36 @@ <h1 itemprop="headline" class="content__headline js-score">@Html(article.trail.h
...
+                                @* Remove LiveFilter and the fragment once Toast comes out of AB test status *@
+                                @fragments.liveBlogNavigation(article.content.metadata.id, model.currentPage)
+                            </div>
+
+                            @fragments.submeta(article)
+
+                            <div class="after-article js-after-article"></div>
+
+                            <div class="js-bottom-marker"></div>
+                        }
```
Comment (TBonnin) [inline article/app/views/liveblog/liveBlogBody.scala.html]: There is a lot of code duplication in this `if/else`. Is there any way to avoid it?

Comment (SiAdcock) [inline article/app/views/liveblog/liveBlogBody.scala.html]: Good shout. I could probably extract everything inside the dropdown into a fragment

**Comment by johnduffell** [comment]
great start! 👍 
@NataliaLKB and I had a few debates about whether it's best to pass the amp boolean around or use the requestHeader implicit when things are really nested.  I think we do different thing depending what feels most appropriate for that code.
Also I'd watch out for having big if (amp) {.......} blocks, I like that you split one of the views out to avoid that.

**Comment by johnduffell** [comment]
I just noticed this isn't behind a switch or opt-in, be careful what google (and twitter uses amp) does, we don't want to accidentally send everyone who googles to our amp version if it doesn't provide the functionality we would expect.

**Comment by SiAdcock** [comment]
Thanks @johnduffell, this is really helpful!

Passing that `amp` Boolean around felt a bit smelly. I shall have another pass and see if I can use `requestHeader.isAmp` directly anyway.

From what I have read, Google will not index the AMP version of the live blog as we do not add an `amphtml` link element to the canonical (non-AMP) version of the live blog. Therefore the AMP version [will not be discoverable by Google](https://support.google.com/webmasters/answer/6340290?hl=en#discovery). However, I can wrap this in a Feature Switch if you think it will be safer. It would also give us greater control, and enable us to quickly turn off the AMP live blog in the event of some catastrophe.

**Comment by TBonnin** [comment]
+1 for a switch. It is easier to manage the release with a `off` default value and safer/quicker to disable if something goes wrong (even though nothing goes wrong ever of course 😬 )

**Comment by SiAdcock** [comment]
After a conversation with @sndrs, I'm going to take another look at the CSS to remove all nested class selectors. This is because we would like the AMP live blog to eventually use [postcss-atomised](https://github.com/sndrs/postcss-atomised), which will not support nested classes for the time being.

**Comment by SiAdcock** [comment]
I've made the following changes / excuses:

-  Added a feature switch for AMP live blogs. Prevents the `amphtml` link tag from being added to the canonical page while the switch is off. Switch is off by default
- De-nestified CSS in preparation for postcss-atomised
- The way the `amp` flag is being passed down the view tree is appropriate in all instances, as it is always being used within the views to determine which elements to show. If the flag was being passed to a view simply to allow that view to pass it to a child, in this instance referring to the `isAmp` flag on the `requestHeader` in the child view would be more appropriate. Happy to hear any views on this.

File: article/app/views/liveblog/liveBlogBodyContent.scala.html [hunk 2]
```diff
@@ -0,0 +1,44 @@
+@import _root_.liveblog.KeyEventData
+@import model.Badges.badgeFor
+
+@import _root_.liveblog.Canonical
```
Comment (johnduffell) [inline article/app/views/liveblog/liveBlogBodyContent.scala.html]: it probably wasn't yours originally, but the line spacing is a little wonky - and actually there's no need for any of the imports in the signature by the looks, so they could all be below (or all above if you prefer that way)

File: common/app/conf/switches/FeatureSwitches.scala [hunk 3]
```diff
@@ -377,6 +377,16 @@ trait FeatureSwitches {
...
   )
 
+  val LiveBlogAmpSwitch = Switch(
+    SwitchGroup.Feature,
+    "live-blog-amp-switch",
+    "If this switch is on, link to amp pages will be in the metadata for live blogs",
+    owners = Seq(Owner.withGithub("SiAdcock")),
+    safeState = Off,
+    sellByDate = never,
+    exposeClientSide = true
```
Comment (johnduffell) [inline common/app/conf/switches/FeatureSwitches.scala]: is it really needed client side? also normally we don't use "switch" in the switch name, just in the name of the val itself

Comment (johnduffell) [inline common/app/conf/switches/FeatureSwitches.scala]: also it should have a sell by date, I don't think this switch will be around for ever

**Comment by johnduffell** [comment]
looks great, just a couple of comments on the switch and then it can go as far as I'm concerned!  I edited the title to make it easier to read too if that's ok! 👓

**Comment by SiAdcock** [comment]
Thanks for all your suggestions @johnduffell, I agree with all of them. I'll remove unnecessary imports, mark the switch for expiry (or review) one month from today, fix the switch name and prevent its exposure on the client.

Title update looks good too 😄

**Comment by johnduffell** [comment]
👍

**Pull request merged**